pkg/guardian/redis_conf_store.go

```diff
@@ -502,5 +745,79 @@ func (rs *RedisConfStore) pipelinedFetchConf() fetchConf {
 		rs.logger.WithError(err).Warnf("error sending HGETALL for key %v", redisRouteRateLimitsEnabledKey)
 	}
 
+	if jailLimitCounts, err := jailLimitCountCmd.Result(); err == nil {
+		for route, countStr := range jailLimitCounts {
+			parsedURL, urlParseErr := url.Parse(route)
+			count, intParseErr := strconv.ParseUint(countStr, 10, 64)
+			if urlParseErr != nil || intParseErr != nil {
+				rs.logger.WithError(urlParseErr).WithError(intParseErr).Warnf("error parsing jail limit count for %v", route)
+			} else {
+				newConf.jailLimitCounts[*parsedURL] = &count
+			}
+		}
+	} else {
+		rs.logger.WithError(err).Warnf("error sending HGETALL for key %v", redisJailLimitsCountKey)
+	}
+
+	if jailLimitDurations, err := jailLimitDurationCmd.Result(); err == nil {
+		for route, durationStr := range jailLimitDurations {
+			parsedURL, urlParseErr := url.Parse(route)
+			duration, durationParseErr := time.ParseDuration(durationStr)
+			if urlParseErr != nil || durationParseErr != nil {
+				rs.logger.WithError(urlParseErr).WithError(durationParseErr).Warnf("error parsing jail limit duration for %v", route)
+			} else {
+				newConf.jailLimitDurations[*parsedURL] = &duration
+			}
+		}
+	} else {
+		rs.logger.WithError(err).Warnf("error sending HGETALL for key %v", redisJailLimitsDurationKey)
+	}
+
+	if jailLimitsEnabled, err := jailLimitEnabledCmd.Result(); err == nil {
+		for route, enabled := range jailLimitsEnabled {
+			parsedURL, urlParseErr := url.Parse(route)
+			enabled, boolParseErr := strconv.ParseBool(enabled)
+			if urlParseErr != nil || boolParseErr != nil {
+				rs.logger.WithError(urlParseErr).WithError(boolParseErr).Warnf("error parsing route limit enabled for %v", route)
+			} else {
+				newConf.jailLimitsEnabled[*parsedURL] = &enabled
+			}
+		}
+	} else {
+		rs.logger.WithError(err).Warnf("error sending HGETALL for key %v", redisJailLimitsEnabledKey)
+	}
+
+	if jailBanDurations, err := jailBanDurationCmd.Result(); err == nil {
+		for route, durationStr := range jailBanDurations {
+			parsedURL, urlParseErr := url.Parse(route)
+			duration, durationParseErr := time.ParseDuration(durationStr)
+			if urlParseErr != nil || durationParseErr != nil {
+				rs.logger.WithError(urlParseErr).WithError(durationParseErr).Warnf("error parsing jail ban duration for %v", route)
+			} else {
+				newConf.jailBanDuration[*parsedURL] = &duration
+			}
+		}
+	} else {
+		rs.logger.WithError(err).Warnf("error sending HGETALL for key %v", redisJailLimitsEnabledKey)
+	}
+
+	if prisoners, err := prisonersCmd.Result(); err == nil {
+		for ip, expiry := range prisoners {
+			expireTime, err := time.Parse("2006-01-02 15:04:05.999999999 -0700 MST", expiry)
+			if err != nil {
+				rs.logger.WithError(err).Warnf("error parsing time (%v): %v", expiry, err)
+			}
+			// only place in map if expire time is later than now, rely upon separate clean up job
+			// to clean up expired bans.
```

Comment:
When is the separate clean up job running? Might we just prune prisoners here?

Reply:
I was picturing this separate clean up job as a cronjob that runs periodically. It would basically just run `guardian-cli prune-prisoners` at an interval that is < the `min(ban_duration)` across all of the jails. 

I was thinking about pruning prisoners here as well, but there are some race conditions with Guardian that made me think that perhaps a separate cronjob is better. 

Guardian doesn't do any kind of locking when updating keys in Redis, and this PR continues with this behavior with the new Jails feature. This leads to some subtle race conditions that aren't really an issue for most of Guardian's features. 

For the case of jails/prisoners, all Guardian instances could add the same prisoner at the same time. However, the risk around this race condition is just that the expiration of the prisoner might get set a few times until all Guardian instances become consistent and agree that the client is a prisoner (should be ~30 seconds). 

In the case of removing a prisoner, the risk of the race condition is that multiple Guardian instances could be adding and removing prisoners at the same time. So Guardian could mistakenly add or remove prisoners. 

By having a single process (e.g. a separately executed cronjob) that handles the removal of all expired prisoners in an atomic fashion, I believe the risk is removed in most cases. 

The alternative, could be to just lock the prisoners key in redis while doing this part of the operation. Perhaps that makes more sense? I was a bit worried about lock contention but given that Guardian only updates its configuration every 30 seconds, it shouldn't really be an issue.

Reply:
> The alternative, could be to just lock the prisoners key in redis while doing this part of the operation.

My concern with an external cronjob is that, without locking, it's also not really race-free and introduces additional operational complexity. How would operators be alerted if the cronjob stopped running for some reason, or failed?

Reply:
Yeah, this does add a lot of operational overhead. Let me try to just add a lock around this portion so we can safely remove the expired prisoners from the guardian instances directly.

As I was writing this comment out, I began to realize that perhaps my concerns around the lock contention were a bit premature and unwarranted.